src/legacy/core_plugins/table_vis/migrations.js

```diff
@@ -17,11 +17,55 @@
  * under the License.
  */
 
-import { cloneDeep, get, omit } from 'lodash';
+import { cloneDeep, get, set, omit } from 'lodash';
+
+function migrateIndexPattern(type, doc) {
+  const searchSourceJSON = get(doc, 'attributes.kibanaSavedObjectMeta.searchSourceJSON');
+  if (typeof searchSourceJSON !== 'string') {
+    return;
+  }
+  let searchSource;
+  try {
+    searchSource = JSON.parse(searchSourceJSON);
+  } catch (e) {
+    // Let it go, the data is invalid and we'll leave it as is
+    return;
+  }
+  if (!searchSource.index) {
+    return;
+  }
+  doc.references.push({
+    name: 'kibanaSavedObjectMeta.searchSourceJSON.index',
+    type: 'index-pattern',
+    id: searchSource.index,
+  });
+  searchSource.indexRefName = 'kibanaSavedObjectMeta.searchSourceJSON.index';
+  delete searchSource.index;
+  set(doc, 'attributes.kibanaSavedObjectMeta.searchSourceJSON', JSON.stringify(searchSource));
+}
 
 export const migrations = {
   visualization: {
     '7.0.0': (doc) => {
+      // Set new "references" attribute
```

Comment:
Lets move this migration into the kibana plugin. It owns the mapping of the visualization saved objects and should also own the migration history of those objects.